src/admin-page/UsersTable.js

```diff
@@ -0,0 +1,245 @@
+  const [password, setPassword] = useState('');
+  const [role, setRole] = useState('');
+  const [firstName, setFirstName] = useState('');
+  const [lastName, setLastName] = useState('');
+
+  const handleOpenUpdateModal = (id) => {
+    setIdToUpdate(id);
+    setOpenUpdateModal(true);
+  };
+
+  const handleOpenDeleteModal = (id) => {
+    setIdToDelete(id);
+    setOpenDeleteModal(true);
+  };
+
+  const handleCloseUpdateModal = () => {
+    setIdToUpdate(null);
+    setOpenUpdateModal(false);
+  };
+
+  const handleCloseDeleteModal = () => {
+    setIdToDelete(null);
+    setOpenDeleteModal(false);
+  };
+
+  const deleteUser = () => {
+    deleteUserMutation({
+      variables: { id: idToDelete },
+      update(cache) {
+        cache.modify({
+          fields: {
+            users(existingUsers, { readField }) {
+              return existingUsers.filter(
+                (userRef) => idToDelete !== readField('id', userRef),
+              );
+            },
+          },
+        });
+      },
+    });
+  };
+
+  const updateUser = (user) => {
+    setOpenUpdateModal(false);
+    updateUserMutation({
+      variables: {
+        id: user.id,
+        username: user.username,
+        password: user.password,
+        firstName: user.firstName,
+        lastName: user.lastName,
+        userRoleId: getUserRoleId(user.role),
+      },
+    });
+  };
+
+  const addUser = () => {
+    setAddRowButtonClicked(false);
+    const newUser = {
+      username, firstName, lastName, password, userRole: getUserRoleId(role),
+    };
+    createUser({
+      variables: {
+        ...newUser,
+      },
+      optimisticResponse: {
+        __typename: 'Mutation',
+        createUser: {
+          __typename: 'User',
+          username,
+          password,
+          firstName,
+          lastName,
+          userRole: {
+            __typename: 'UserRole',
+            name: role,
+          },
+        },
+      },
+      update(cache, { data: mutationData }) {
+        const userToAdd = {
+          user: {
+            __typename: 'User',
+            id: mutationData.createUser?.id,
+            username: mutationData.createUser?.username,
+            password: mutationData.createUser?.password,
+            firstName: mutationData.createUser?.firstName,
+            lastName: mutationData.createUser?.lastName,
+            userRole: {
+              __typename: 'UserRole',
+              name: mutationData.createUser?.userRole?.name,
+            },
+          },
+        };
+        try {
+          cache.writeQuery({
+            query: GET_USERS,
+            data: {
+              users: [userToAdd.user, ...data?.users],
+            },
+          });
+        } catch (e) {
+          cache.writeQuery({
+            query: GET_USERS,
+            data: {
+              users: [userToAdd.user],
+            },
+          });
+        }
+      },
+    });
+  };
+
+  return (
+    <>
+
+      <Modal
```

Comment:
This component is kind of big, some things could be extracted. The modals for example.

Reply:
Extracted some, a lot of it's content are the functions manipulating the cache.